qualtran/surface_code/ccz2t_cost_model.py

```diff
@@ -204,11 +225,60 @@ def get_ccz2t_costs(
         data_d = code_distance_from_budget(phys_err=phys_err, budget=target_err_per_round)
         data_block = SimpleDataBlock(data_d=data_d, routing_overhead=routing_overhead)
 
-    data_error = data_block.data_error(
-        n_algo_qubits=n_algo_qubits, n_cycles=n_cycles, phys_err=phys_err
+    return get_ccz2t_costs(
+        n_magic=n_magic,
+        n_algo_qubits=n_algo_qubits,
+        phys_err=phys_err,
+        cycle_time_us=cycle_time_us,
+        factory=factory,
+        data_block=data_block,
     )
-    failure_prob = distillation_error + data_error
-    footprint = factory.footprint() + data_block.footprint(n_algo_qubits=n_algo_qubits)
-    duration_hr = (cycle_time_us * n_cycles) / (1_000_000 * 60 * 60)
 
-    return PhysicalCost(failure_prob=failure_prob, footprint=footprint, duration_hr=duration_hr)
+
+def iter_ccz2t_factories(
+    l1_start: int = 5, l1_stop: int = 25, l2_stop: int = 41
+) -> Iterator[CCZ2TFactory]:
+    for l1_distance in range(l1_start, l1_stop, 2):
+        for l2_distance in range(l1_distance + 2, l2_stop, 2):
+            yield CCZ2TFactory(distillation_l1_d=l1_distance, distillation_l2_d=l2_distance)
+
+
+def iter_simple_data_blocks(d_start: int = 7, d_stop: int = 35):
+    for logical_data_qubit_distance in range(d_start, d_stop, 2):
+        yield SimpleDataBlock(data_d=logical_data_qubit_distance)
+
+
+def get_ccz2t_costs_from_grid_search(
+    *,
+    n_magic: AlgorithmSummary,
+    n_algo_qubits: int,
+    phys_err: float = 1e-3,
+    cycle_time_us: float = 1.0,
+    factory_iter: Callable[[], Iterator] = iter_ccz2t_factories,
```

Comment:
Why do we request a generator factory_iter: Callable[[], Iterator] = iter_ccz2t_factories rather than an iterator factory_iter: Iterator = iter_ccz2t_factories()?

Reply:
that could work. You'd have to change the way we set the default value since I _think_ calling the generator in the default argument would exhibit some really tricky and unwanted behavior

Reply:
What about allowing both an iterator and a generator? 
i.e. 
```
def get_ccz2t_costs_from_grid_search(
    [...],
    factory_iter: Union[Callable[[], Iterator], Iterator] = iter_ccz2t_factories,
    [...]):
    if isinstance(iter_ccz2t_factories, Callable):
        iter_ccz2t_factories = iter_ccz2t_factories()
```

Reply:
If we like this (or another) solution, I can add it in #561 before finalizing

Reply:
Feel free to change it to (just) an iterator. Just make sure the defaults work!

Reply:
I think `Iterable` is most suitable here since it covers lists, tuples and iterators. The default value should be be `tuple(iter_ccz2t_factories())` however since the default value gets initialized once so the iterator will get consumed during the first call